### PR TITLE
8305252: make_method_handle_intrinsic may call java code under a lock

### PR DESCRIPTION
**Pull request by coleenp**
This patch releases the InvokeMethodTable_lock while creating a method handle intrinsic.  If there's a race, it frees a Method created by racing thread.  The logic is simple but uses the deallocate_list infrastructure that's mostly used for redefinition making it less rare.  With Dacapo2009, this adds about 20 Methods + constant pools to the list.  Also the method has to call nmethod->flush which is assumed to be something only GC calls.

Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305252](https://bugs.openjdk.org/browse/JDK-8305252): make_method_handle_intrinsic may call java code under a lock

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [ed40ee56](https://git.openjdk.org/jdk/pull/13308/files/ed40ee56412693f1fb6ed39bed1e02b39213b1e5)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13308/head:pull/13308` \
`$ git checkout pull/13308`

Update a local copy of the PR: \
`$ git checkout pull/13308` \
`$ git pull https://git.openjdk.org/jdk.git pull/13308/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13308`

View PR using the GUI difftool: \
`$ git pr show -t 13308`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13308.diff">https://git.openjdk.org/jdk/pull/13308.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13308#issuecomment-1513368416)

### CODE REVIEW

**Review by iklam — COMMENTED** [review]
Looks reasonable to me.

**Review by iklam — APPROVED** [review]

**Review by dholmes-ora — COMMENTED** [review]
>  Also the method has to call nmethod->flush which is assumed to be something only GC calls.

I'm not familiar with this code but this raises some concerns for me. Is there a way we can test this deallocation path?

Other comments below.

Thanks.

File: src/hotspot/share/oops/method.cpp [hunk 1]
```diff
@@ -135,8 +135,12 @@ void Method::deallocate_contents(ClassLoaderData* loader_data) {
   set_method_data(nullptr);
   MetadataFactory::free_metadata(loader_data, method_counters());
   clear_method_counters();
-  // The nmethod will be gone when we get here.
-  if (code() != nullptr) _code = nullptr;
+  // The nmethod will be gone when we get here, for redefinition but not
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/method.cpp]: I updated the comment.  Regular class unloading does not get here.  The Metaspace containing the Method is released, not requiring individual metadata deallocation.
This is tested with normal execution when the race to create method handle intrinsic Method is lost.  I counted ~20 with logging for Dacapo.

File: src/hotspot/share/oops/method.cpp [hunk 2]
```diff
@@ -135,8 +135,12 @@ void Method::deallocate_contents(ClassLoaderData* loader_data) {
...
+  // for method handle intrinsics.
+  if (code() != nullptr) {
+    ((nmethod*)_code)->flush();
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/method.cpp:141]: The call path to this is during a safepoint for cleaning up the deallocate lists, and the CodeCache_lock is a pretty low level no-safepoint lock, so the lock is fine.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the updated comments and answering my queries. Looks fine.

**Comment by coleenp** [comment]
Talking with @fisk makes me think the nmethod->flush is unsafe in the long term.  The nmethod->flush takes out the CodeCache_lock, but deoptimization uses a RelaxedCompiledMethodIterator to iterate over nmethods.  It only takes out the CodeCache_lock when calling next().  It assumes that the nmethod it returns won't be concurrently destroyed by GC, but this code also destroys an nmethod.

This code is only called during a dedicated safepoint so is safe from deoptimization for now.

But ClassLoaderDataGraph::purge can be called concurrently.  CLDG::purge doesn't call Method::deallocate_contents because the metadata is released in bulk, but if this intrinsic Method is on the deallocate list we call free_deallocate_list_C_heap_structures() which doesn't flush the nmethod for this Method.  So an nmethod in the code cache would point to a deleted Method.  But it can't call flush() because this isn't in a safepoint and violate the code cache walking assumptions.

The only ClassLoaderData that have these Methods on the deallocate_lists are ones that are never unloaded so this is safe for now.

The "for now" safety is why I'm closing this PR and going with a different approach.

**Pull request closed without merging**